File: src/java_answer/Lessons12945.java
package java_answer;

import java.math.BigInteger;
import java.util.HashMap;
import java.util.Map;

/**
 * @author dave.th
 * https://programmers.co.kr/learn/courses/30/lessons/12945
 */
public class Lessons12945 {
    public static void main(String[] args) {
        System.out.println((new Lessons12945()).solution(10000));
    }

    private int solution(int n) {
        Map<Integer, BigInteger> map = new HashMap<>();
        map.put(0, BigInteger.valueOf(0));
        map.put(1, BigInteger.valueOf(1));

        for(int i=0; i <= n; i++)
           fibo(i, map);

        return Integer.valueOf(map.get(n).mod(BigInteger.valueOf(1234567)).toString());
    }

    private BigInteger fibo(int n, Map<Integer, BigInteger> map) {
        if(n == 0 || n == 1) return BigInteger.valueOf(n);
        if(map.containsKey(n)) return map.get(n);

        BigInteger result = fibo(n - 1, map).add(fibo(n - 2, map));
        map.put(n, result);
        return result;
    }
}

File: src/scala_answer/Lessons12945.scala
package scala_answer

/**
  * @author dave.th
  * https://programmers.co.kr/learn/courses/30/lessons/12945
  */
object Lessons12945 {
  def main(args: Array[String]) : Unit = {
    println(solution(10000))
  }

  def solution(n : Int) : Int = {
    val map = scala.collection.mutable.Map[Int, BigInt]()
    map.put(0, 0)
    map.put(1, 1)

    for(i <- 0 to n)
      fibo(i, map)

    map.getOrElse(n, BigInt(0)).mod(1234567).toInt
  }

  def fibo(n : Int, map : scala.collection.mutable.Map[Int, BigInt]) : BigInt = {
    if(n == 0 || n == 1) return n

    val mapResult : BigInt = map.getOrElse(n, -1)
    if(mapResult > -1) return mapResult

    val result = fibo(n - 1, map) + fibo(n - 2, map)
    map.put(n, result)
    result
  }
}
